apiserver/provider/provider_test.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"github.com/juju/errors"
+	jujutesting "github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	names "gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/apiserver/testing"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/provider/dummy"
+	"github.com/juju/juju/state"
+)
+
+type providerTypesSuite struct{}
+
+var _ = gc.Suite(&providerTypesSuite{})
+
+var over9kCPUCores uint64 = 9001
+
+func (p *providerTypesSuite) TestInstanceTypes(c *gc.C) {
+	backend := mockBackend{}
+	authorizer := testing.FakeAuthorizer{Tag: names.NewUserTag("admin"),
+		EnvironManager: true}
+	itCons := constraints.Value{CpuCores: &over9kCPUCores}
+	failureCons := constraints.Value{}
+	m := mockEnviron{
+		results: map[constraints.Value]instances.InstanceTypesWithCostMetadata{
+			itCons: instances.InstanceTypesWithCostMetadata{
+				CostUnit:     "USD/h",
```

Comment:
It looks weird having both USD/h and USD, but let's come back to it after seeing exactly how the GUI wants to format things. Easy enough to massage into shape.